File: tests/oracle/rule_subcommute_freevn_1204.cc
/*
 * RULE-SUBCOMMUTE-FREEVN-0001: locked Ghidra 12.0.4 fixture.
 *
 * Drives RuleSubCommute::applyOp (ruleaction.cc:4514-4653) directly on
 * synthetic SUBPIECE( {INT_ADD,INT_MULT,INT_AND,INT_OR,INT_XOR,INT_NEGATE}(
 * freeVn / const, ... ), offset ) forms — the generic tail commute loop at
 * ruleaction.cc:4631-4652 whose opSetInput ORDER carries the semantics this
 * ticket locks (BINSWEEP-SUBCOMMUTE-FREEVARNODE-0001, CR-SUBCOMMUTE finding
 * 3): cc:4640 opSetInput(longform,newVn,i) must free vn from longform's slot
 * i BEFORE cc:4641 opSetInput(newsub,vn,0) attaches it, because a free
 * varnode may hold only one descendant (varnode.cc:330-340 addDescend throws
 * "Free varnode has multiple descendants" on the second) and a constant
 * still read elsewhere takes the Funcdata::opSetInput dedup COPY
 * (funcdata_op.cc:108-115) instead of the original varnode identity.
 *
 * Free varnodes are manufactured exactly the way the real pipeline does:
 * a COPY writes a unique varnode, longform reads it, then
 * Funcdata::opUnsetOutput(writer) (funcdata_op.cc:52-66 -> VarnodeBank::
 * makeFree, varnode.cc:1316) clears WRITTEN while keeping the descend list.
 *
 * Modes (argv[1]):
 *   normal     — 13 cases, golden stdout diff vs the Rust mirror.
 *   trap_dupfree — one case where the SAME free varnode feeds both XOR
 *                slots (legal while written; freed afterwards): the oracle
 *                itself throws LowlevelError rc=1 when the tail's first
 *                slot attach finds the leftover second descend entry
 *                (varnode.cc:334-336). Form comparison only, no golden —
 *                the crash-form pair is the deliverable.
 *
 * Observations (one block per case, ops in SeqNum order):
 *   case=<name>|sub_apply=<0/1>
 *     op=<opcode#>@0x<addr>|nin=<k>|in0=<c|w|o>|<in0_size>|0x<in0_off>|out=<n>
 *     keep=<c0|c1|f0|f1|w0|w1>|readers=<n>|descends=<n>
 *   endcase
 * in0 class: c=constant, w=written (has def), o=other (free).
 * keep lines: per captured longform input (slot order), readers = alive
 * SUBPIECE ops in the case window whose in(0) IS the captured Varnode
 * (pointer identity — catches the dedup-copy deviation), descends = live
 * descendant count. Each case owns the address window [base, base+0x80);
 * rule-created ops inherit the SUBPIECE address (data.newOp(2,
 * op->getAddr()), cc:4637).
 */

#include "bfd_arch.hh"
#include "libdecomp.hh"
#include "ruleaction.hh"

#include <iostream>
#include <sstream>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::ostringstream;
using std::string;
using std::vector;

RuleSubCommute subCommuteRule("analysis");

Varnode *constantInput(Funcdata &fd, uintb value, int4 size)
{
  return fd.newConstant(size, value);
}

void printOpLine(PcodeOp *op)
{
  ostringstream out;
  out << "  op=" << static_cast<int4>(op->code())
      << "@0x" << std::hex << op->getAddr().getOffset() << std::dec
      << "|nin=" << op->numInput();
  Varnode *in0 = op->numInput() > 0 ? op->getIn(0) : (Varnode *)0;
  if (in0 != (Varnode *)0) {
    char cls = 'o';
    if (in0->isConstant())
      cls = 'c';
    else if (in0->isWritten())
      cls = 'w';
    out << "|in0=" << cls
        << "|" << in0->getSize()
        << "|0x" << std::hex << in0->getOffset() << std::dec;
  }
  else
    out << "|in0=_|0|0x0";
  out << "|out=" << (op->getOut() != (Varnode *)0 ? op->getOut()->getSize() : -1)
      << '\n';
  std::cout << out.str();
}

void dumpCaseWindow(Funcdata &fd, const Address &lo, const Address &hi)
{
  for (auto iter = fd.beginOpAll(); iter != fd.endOpAll(); ++iter) {
    PcodeOp *op = iter->second;
    const Address &a = op->getAddr();
    if (a < lo || hi < a)
      continue;
    printOpLine(op);
  }
}

void dumpKeepLines(Funcdata &fd, const vector<Varnode *> &keeps,
                   const vector<string> &tags, const Address &lo,
                   const Address &hi)
{
  for (uint4 k = 0; k < keeps.size(); ++k) {
    Varnode *vn = keeps[k];
    int4 readers = 0;
    for (auto iter = fd.beginOpAll(); iter != fd.endOpAll(); ++iter) {
      PcodeOp *op = iter->second;
      const Address &a = op->getAddr();
      if (a < lo || hi < a)
        continue;
      if (op->code() != CPUI_SUBPIECE)
        continue;
      if (op->numInput() > 0 && op->getIn(0) == vn)
        ++readers;
    }
    int4 descends = 0;
    for (auto iter = vn->beginDescend(); iter != vn->endDescend(); ++iter)
      ++descends;
    std::cout << "  keep=" << tags[k] << "|readers=" << readers
              << "|descends=" << descends << '\n';
  }
}

// ---- case builder ----------------------------------------------------------

// slotKind per longform slot: 0=free (COPY writer, unset after wiring),
// 1=constant, 4=written (COPY writer kept alive), 3=same vn as slot 0
// freed afterwards (trap_dupfree only), 5=same vn as slot 0 kept written
// (dup-reuse leg). extraKind: 0=none, 1=ZEXT overlap reader on outvn
// (cc:4623-4628 reject), 2=second SUBPIECE reader on base (cc:4621 reject).
struct CaseBuilt {
  PcodeOp *subOp;
  vector<Varnode *> keeps;
  vector<string> tags;
};

CaseBuilt tailCommuteCase(Funcdata &fd, BlockBasic *block, AddrSpace *code,
                          OpCode opc, uint4 base, const int4 *slotKind,
                          const uintb *slotVal, int4 longSize, int4 outSize,
                          int4 offset, int4 extraKind)
{
  CaseBuilt r;
  r.subOp = (PcodeOp *)0;
  int4 nin = (opc == CPUI_INT_NEGATE) ? 1 : 2;
  Varnode *ins[2] = {(Varnode *)0, (Varnode *)0};
  PcodeOp *writers[2] = {(PcodeOp *)0, (PcodeOp *)0};
  for (int4 slot = 0; slot < nin; ++slot) {
    if (slotKind[slot] == 3 || slotKind[slot] == 5) {
      ins[slot] = ins[0];
      continue;
    }
    if (slotKind[slot] == 1) {
      ins[slot] = constantInput(fd, slotVal[slot], longSize);
      continue;
    }
    PcodeOp *writer = fd.newOp(1, Address(code, base + 0x10 + 0x8 * slot));
    fd.opSetOpcode(writer, CPUI_COPY);
    Varnode *fv = fd.newUniqueOut(longSize, writer);
    fd.opSetInput(writer, constantInput(fd, slotVal[slot], longSize), 0);
    fd.opInsertEnd(writer, block);
    writers[slot] = writer;
    ins[slot] = fv;
  }
  PcodeOp *longform = fd.newOp(nin, Address(code, base + 0x30));
  fd.opSetOpcode(longform, opc);
  fd.newUniqueOut(longSize, longform);
  fd.opSetInput(longform, ins[0], 0);
  if (nin > 1)
    fd.opSetInput(longform, ins[1], 1);
  fd.opInsertEnd(longform, block);
  PcodeOp *subOp = fd.newOp(2, Address(code, base + 0x40));
  fd.opSetOpcode(subOp, CPUI_SUBPIECE);
  fd.newUniqueOut(outSize, subOp);
  fd.opSetInput(subOp, longform->getOut(), 0);
  fd.opSetInput(subOp, constantInput(fd, offset, 4), 1);
  fd.opInsertEnd(subOp, block);
  // Free the COPY outputs: fv keeps descend=[longform(,longform)] but loses
  // WRITTEN — the exact free-varnode state the ip corpus hit.
  for (int4 slot = 0; slot < nin; ++slot) {
    if (slotKind[slot] == 0)
      fd.opUnsetOutput(writers[slot]);
  }
  if (extraKind == 1) {
    // outvn->loneDescend() is a ZEXT back to insize: cc:4626-4627 reject.
    PcodeOp *zx = fd.newOp(1, Address(code, base + 0x50));
    fd.opSetOpcode(zx, CPUI_INT_ZEXT);
    fd.newUniqueOut(longSize, zx);
    fd.opSetInput(zx, subOp->getOut(), 0);
    fd.opInsertEnd(zx, block);
  }
  else if (extraKind == 2) {
    // A second SUBPIECE reads base: base->loneDescend() != op reject.
    PcodeOp *sub2 = fd.newOp(2, Address(code, base + 0x50));
    fd.opSetOpcode(sub2, CPUI_SUBPIECE);
    fd.newUniqueOut(outSize, sub2);
    fd.opSetInput(sub2, longform->getOut(), 0);
    fd.opSetInput(sub2, constantInput(fd, offset, 4), 1);
    fd.opInsertEnd(sub2, block);
  }
  for (int4 slot = 0; slot < nin; ++slot) {
    r.keeps.push_back(ins[slot]);
    ostringstream tag;
    tag << ((slotKind[slot] == 1)   ? 'c'
            : ((slotKind[slot] == 4 || slotKind[slot] == 5) ? 'w' : 'f'))
        << slot;
    r.tags.push_back(tag.str());
  }
  r.subOp = subOp;
  return r;
}

struct CaseSpec {
  const char *name;
  uint4 base;
  int opc; // 0 ADD, 1 MULT, 2 AND, 3 OR, 4 XOR, 5 NEGATE
  int4 slotKind[2];
  uintb slotVal[2];
  int4 longSize;
  int4 outSize;
  int4 offset;
  int4 extraKind;
};

} // anonymous namespace

int main(int argc, char **argv)
{
  if (argc != 4) {
    std::cerr << "usage: rule_subcommute_freevn_1204 SPEC_ROOT CURL_BINARY MODE(normal|trap_dupfree)\n";
    return 2;
  }
  try {
    vector<string> specPaths;
    specPaths.push_back(argv[1]);
    startDecompilerLibrary(specPaths);
    {
      std::ostringstream diagnostics;
      BfdArchitecture architecture(argv[2], "default", &diagnostics);
      DocumentStorage store;
      architecture.init(store);
      architecture.readLoaderSymbols("::");
      Funcdata *fd = architecture.symboltab->getGlobalScope()->queryFunction("GetStr");
      if (fd == (Funcdata *)0)
        throw std::runtime_error("GetStr was not found in the BFD symbol table");
      if (fd->getName() != "GetStr" ||
          fd->getAddress().getOffset() != 0x36d0 || fd->getSize() != 0)
        throw std::runtime_error("GetStr input identity drifted");
      if (architecture.archid != "x86:LE:64:default:gcc")
        throw std::runtime_error("runtime architecture/compiler drifted: " +
                                 architecture.archid);

      string mode = argv[3];
      AddrSpace *code = fd->getArch()->getDefaultCodeSpace();
      BlockGraph &graph = const_cast<BlockGraph &>(fd->getBasicBlocks());
      BlockBasic *block = graph.newBlockBasic(fd);

      const vector<CaseSpec> cases = {
        // The ip-corpus panic form: both longform inputs free varnodes.
        {"add_free_free", 0x510000, 0, {0, 0}, {0x11, 0x22}, 8, 4, 0, 0},
        // Const tail inputs: identity must survive (no opSetInput dedup copy
        // once cc:4640 frees the slot first).
        {"add_free_const", 0x510100, 0, {0, 1}, {0x11, 0x64}, 8, 4, 0, 0},
        {"add_const_free", 0x510200, 0, {1, 0}, {0x64, 0x22}, 8, 4, 0, 0},
        {"add_const_const", 0x510300, 0, {1, 1}, {0x64, 0xc8}, 8, 4, 0, 0},
        // Other commuting opcodes with free tails.
        {"mult_free_free", 0x510400, 1, {0, 0}, {7, 6}, 8, 4, 0, 0},
        {"and_free_free", 0x510500, 2, {0, 0}, {0xf0, 0x0f}, 8, 2, 0, 0},
        {"or_free16", 0x510600, 3, {0, 0}, {0xa1, 0xb2}, 16, 8, 0, 0},
        // Single-input arm.
        {"negate_free", 0x510700, 5, {0, 0}, {0x33, 0}, 8, 4, 0, 0},
        // Bitwise commutes at a non-zero subpiece offset.
        {"xor_free_off4", 0x510800, 4, {0, 0}, {0x44, 0x55}, 8, 4, 4, 0},
        {"mult_free_const", 0x510900, 1, {0, 1}, {9, 3}, 8, 4, 0, 0},
        // dup-reuse leg: the SAME written vn feeds both slots — only one
        // new SUBPIECE is created (cc:4636 guard, cc:4646 reuse).
        {"xor_same_written", 0x510a00, 4, {4, 5}, {0x66, 0x66}, 8, 4, 0, 0},
        // cc:4623-4628 reject: outvn feeds a ZEXT back to insize.
        {"add_free_zext_overlap", 0x510b00, 0, {0, 0}, {0x11, 0x22}, 8, 4, 0, 1},
        // cc:4621 reject: a second SUBPIECE also reads base.
        {"add_two_readers", 0x510c00, 0, {0, 0}, {0x11, 0x22}, 8, 4, 0, 2},
      };
      vector<CaseSpec> runCases = cases;
      if (mode == "trap_dupfree") {
        CaseSpec trap = {"trap_dupfree", 0x520000, 4, {0, 3},
                         {0x77, 0x77}, 8, 4, 0, 0};
        runCases.clear();
        runCases.push_back(trap);
      }
      else if (mode != "normal") {
        throw std::runtime_error("unknown mode: " + mode);
      }

      for (uint4 ci = 0; ci < runCases.size(); ++ci) {
        const CaseSpec &spec = runCases[ci];
        OpCode opc = (spec.opc == 0)   ? CPUI_INT_ADD
                     : (spec.opc == 1)  ? CPUI_INT_MULT
                     : (spec.opc == 2)  ? CPUI_INT_AND
                     : (spec.opc == 3)  ? CPUI_INT_OR
                     : (spec.opc == 4)  ? CPUI_INT_XOR
                                        : CPUI_INT_NEGATE;
        CaseBuilt built = tailCommuteCase(*fd, block, code, opc, spec.base,
                                          spec.slotKind, spec.slotVal,
                                          spec.longSize, spec.outSize,
                                          spec.offset, spec.extraKind);
        int4 apply = subCommuteRule.applyOp(built.subOp, *fd);
        std::cout << "case=" << spec.name << "|sub_apply=" << apply << '\n';
        dumpCaseWindow(*fd, Address(code, spec.base),
                       Address(code, spec.base + 0x80));
        dumpKeepLines(*fd, built.keeps, built.tags, Address(code, spec.base),
                      Address(code, spec.base + 0x80));
        std::cout << "endcase\n";
        std::cout.flush();
      }
    }
    shutdownDecompilerLibrary();
    return 0;
  }
  catch(const ghidra::LowlevelError &error) {
    std::cerr << "Ghidra LowlevelError: " << error.explain << '\n';
  }
  catch(const ghidra::DecoderError &error) {
    std::cerr << "Ghidra DecoderError: " << error.explain << '\n';
  }
  catch(const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
  }
  return 1;
}

File: tests/oracle/rule_subcommute_freevn_1204.rs
//! RULE-SUBCOMMUTE-FREEVN-0001: Rugra side of the locked 12.0.4 oracle
//! fixture for the RuleSubCommute generic tail commute loop
//! (ruleaction.cc:4631-4652) on free-varnode and constant inputs
//! (BINSWEEP-SUBCOMMUTE-FREEVARNODE-0001, CR-SUBCOMMUTE finding 3).
//!
//! Mirrors `rule_subcommute_freevn_1204.cc` case-for-case (same names,
//! same observation format):
//!   case=<name>|sub_apply=<0/1>
//!     op=<opcode#>@0x<addr>|nin=<k>|in0=<c|w|o>|<in0_size>|0x<in0_off>|out=<n>
//!     keep=<c0|c1|f0|f1|w0|w1>|readers=<n>|descends=<n>
//!   endcase
//!
//! The opSetInput order under test: cc:4640 frees vn from longform slot i
//! (opSetInput(longform,newVn,i)) BEFORE cc:4641 attaches vn to newsub
//! (opSetInput(newsub,vn,0) — "vn may be free, so set as input after
//! setting newVn"), so a free varnode never gains a second descendant
//! (varnode.cc:330-340) and a constant keeps its identity instead of
//! taking the Funcdata::opSetInput dedup copy (funcdata_op.cc:108-115).
//!
//! Modes: `normal` (golden diff), `trap_dupfree` (same free varnode in
//! both XOR slots: both sides must crash with "Free varnode has multiple
//! descendants" — oracle LowlevelError rc=1 vs Rust panic rc=101 at
//! varnode.rs:2716; form comparison only, no golden).

use std::sync::{Arc, RwLock};

use rugra::action::Rule;
use rugra::address::Address;
use rugra::arch::Architecture;
use rugra::block::{BlockBasic, FlowBlock};
use rugra::funcdata::Funcdata;
use rugra::opcodes::OpCode;
use rugra::ruleaction::RuleSubCommute;
use rugra::varnode::Varnode;

type VnRef = Arc<RwLock<Varnode>>;

fn constant_input(fd: &mut Funcdata, value: u64, size: usize) -> VnRef {
    fd.new_constant(size, value)
}

fn print_op_line(op: &rugra::op::PcodeOpRef) {
    let g = op.0.read().unwrap();
    let addr = g.get_addr().to_space_address().get_offset();
    let mut line = format!("  op={}@0x{:x}|nin={}", g.opcode as i32, addr, g.num_input());
    // Ghidra's destroyed/mislinked slots read as null (op.cc:98 clearInput);
    // Rugra's shared null_slot_sentinel is the same observable — render as `_`.
    let in0 = g.get_in(0).filter(|v| {
        !std::sync::Arc::ptr_eq(*v, &rugra::op::null_slot_sentinel())
    });
    if let Some(in0) = in0 {
        let i = in0.read().unwrap();
        let cls = if i.is_constant() {
            'c'
        } else if i.is_written() {
            'w'
        } else {
            'o'
        };
        line += &format!("|in0={cls}|{}|0x{:x}", i.get_size(), i.get_offset());
    } else {
        line += "|in0=_|0|0x0";
    }
    let out_size = g
        .output
        .as_ref()
        .map(|v| v.read().unwrap().get_size() as i64)
        .unwrap_or(-1);
    line += &format!("|out={out_size}");
    println!("{line}");
}

fn dump_case_window(fd: &Funcdata, lo: u64, hi: u64) {
    for op in fd.begin_op_all() {
        let a = op.0.read().unwrap().get_addr().to_space_address().get_offset();
        if a < lo || hi < a {
            continue;
        }
        print_op_line(op);
    }
}

fn dump_keep_lines(
    fd: &Funcdata,
    keeps: &[VnRef],
    tags: &[String],
    lo: u64,
    hi: u64,
) {
    for (k, vn) in keeps.iter().enumerate() {
        let mut readers = 0;
        for op in fd.begin_op_all() {
            let g = op.0.read().unwrap();
            let a = g.get_addr().to_space_address().get_offset();
            if a < lo || hi < a {
                continue;
            }
            if g.opcode != OpCode::CPUI_SUBPIECE {
                continue;
            }
            if let Some(in0) = g.get_in(0) {
                if std::sync::Arc::ptr_eq(&in0, vn) {
                    readers += 1;
                }
            }
        }
        let descends = vn.read().unwrap().count_descends();
        println!("  keep={}|readers={readers}|descends={descends}", tags[k]);
    }
}

// ---- case builder (mirrors the .cc builder 1:1) ----------------------------

// slotKind per longform slot: 0=free (COPY writer, unset after wiring),
// 1=constant, 4=written (COPY writer kept alive), 3=same vn as slot 0
// freed afterwards (trap_dupfree only), 5=same vn as slot 0 kept written
// (dup-reuse leg). extra_kind: 0=none, 1=ZEXT overlap reader on outvn
// (cc:4623-4628 reject), 2=second SUBPIECE reader on base (cc:4621 reject).
struct CaseBuilt {
    sub_op: rugra::op::PcodeOpRef,
    keeps: Vec<VnRef>,
    tags: Vec<String>,
}

#[allow(clippy::too_many_arguments)]
fn tail_commute_case(
    fd: &mut Funcdata,
    block: &Arc<RwLock<dyn FlowBlock + Send + Sync>>,
    opc: OpCode,
    base: u64,
    slot_kind: [i32; 2],
    slot_val: [u64; 2],
    long_size: usize,
    out_size: usize,
    offset: u64,
    extra_kind: i32,
) -> CaseBuilt {
    let nin = if opc == OpCode::CPUI_INT_NEGATE { 1 } else { 2 };
    let mut ins: Vec<Option<VnRef>> = vec![None, None];
    let mut writers: Vec<Option<rugra::op::PcodeOpRef>> = vec![None, None];
    for slot in 0..nin {
        if slot_kind[slot] == 3 || slot_kind[slot] == 5 {
            ins[slot] = ins[0].clone();
            continue;
        }
        if slot_kind[slot] == 1 {
            ins[slot] = Some(constant_input(fd, slot_val[slot], long_size));
            continue;
        }
        let writer = fd.new_op(1, Address::new(base + 0x10 + 0x8 * slot as u64));
        fd.op_set_opcode(&writer, OpCode::CPUI_COPY);
        let fv = fd.new_unique_out(long_size, &writer);
        let c_arg = constant_input(fd, slot_val[slot], long_size);
        fd.op_set_input(&writer, c_arg, 0);
        fd.op_insert_end(&writer, block);
        writers[slot] = Some(writer);
        ins[slot] = Some(fv);
    }
    let longform = fd.new_op(nin, Address::new(base + 0x30));
    fd.op_set_opcode(&longform, opc);
    fd.new_unique_out(long_size, &longform);
    fd.op_set_input(&longform, ins[0].clone().unwrap(), 0);
    if nin > 1 {
        fd.op_set_input(&longform, ins[1].clone().unwrap(), 1);
    }
    fd.op_insert_end(&longform, block);
    let sub_op = fd.new_op(2, Address::new(base + 0x40));
    fd.op_set_opcode(&sub_op, OpCode::CPUI_SUBPIECE);
    fd.new_unique_out(out_size, &sub_op);
    let addout = longform.0.read().unwrap().output.as_ref().unwrap().clone();
    fd.op_set_input(&sub_op, addout, 0);
    let off_const = constant_input(fd, offset, 4);
    fd.op_set_input(&sub_op, off_const, 1);
    fd.op_insert_end(&sub_op, block);
    // Free the COPY outputs: fv keeps descend=[longform(,longform)] but
    // loses WRITTEN — the exact free-varnode state the ip corpus hit.
    for slot in 0..nin {
        if slot_kind[slot] == 0 {
            fd.op_unset_output(&writers[slot].clone().unwrap());
        }
    }
    if extra_kind == 1 {
        // outvn->loneDescend() is a ZEXT back to insize: cc:4626-4627 reject.
        let zx = fd.new_op(1, Address::new(base + 0x50));
        fd.op_set_opcode(&zx, OpCode::CPUI_INT_ZEXT);
        fd.new_unique_out(long_size, &zx);
        let subout = sub_op.0.read().unwrap().output.as_ref().unwrap().clone();
        fd.op_set_input(&zx, subout, 0);
        fd.op_insert_end(&zx, block);
    } else if extra_kind == 2 {
        // A second SUBPIECE reads base: base->loneDescend() != op reject.
        let sub2 = fd.new_op(2, Address::new(base + 0x50));
        fd.op_set_opcode(&sub2, OpCode::CPUI_SUBPIECE);
        fd.new_unique_out(out_size, &sub2);
        let addout = longform.0.read().unwrap().output.as_ref().unwrap().clone();
        fd.op_set_input(&sub2, addout, 0);
        let off_const = constant_input(fd, offset, 4);
        fd.op_set_input(&sub2, off_const, 1);
        fd.op_insert_end(&sub2, block);
    }
    let mut keeps = Vec::new();
    let mut tags = Vec::new();
    for slot in 0..nin {
        keeps.push(ins[slot].clone().unwrap());
        let ch = match slot_kind[slot] {
            1 => 'c',
            4 | 5 => 'w',
            _ => 'f',
        };
        tags.push(format!("{ch}{slot}"));
    }
    CaseBuilt { sub_op, keeps, tags }
}

struct CaseSpec {
    name: &'static str,
    base: u64,
    opc: i32, // 0 ADD, 1 MULT, 2 AND, 3 OR, 4 XOR, 5 NEGATE
    slot_kind: [i32; 2],
    slot_val: [u64; 2],
    long_size: usize,
    out_size: usize,
    offset: u64,
    extra_kind: i32,
}

fn opc_of(spec: &CaseSpec) -> OpCode {
    match spec.opc {
        0 => OpCode::CPUI_INT_ADD,
        1 => OpCode::CPUI_INT_MULT,
        2 => OpCode::CPUI_INT_AND,
        3 => OpCode::CPUI_INT_OR,
        4 => OpCode::CPUI_INT_XOR,
        _ => OpCode::CPUI_INT_NEGATE,
    }
}

fn run() {
    let mut fd = Funcdata::new("GetStr", Address::new(0x36d0), 0);
    let mut architecture = Architecture::new();
    architecture.archid = "x86:LE:64:default:gcc".to_string();
    fd.set_arch(Arc::new(architecture));
    let block: Arc<RwLock<dyn FlowBlock + Send + Sync>> =
        Arc::new(RwLock::new(BlockBasic::new(0, Address::new(0x5000))));
    fd.bblocks.add_block(block.clone());

    let cases: Vec<CaseSpec> = vec![
        // The ip-corpus panic form: both longform inputs free varnodes.
        CaseSpec { name: "add_free_free", base: 0x510000, opc: 0, slot_kind: [0, 0], slot_val: [0x11, 0x22], long_size: 8, out_size: 4, offset: 0, extra_kind: 0 },
        // Const tail inputs: identity must survive (no opSetInput dedup copy
        // once cc:4640 frees the slot first).
        CaseSpec { name: "add_free_const", base: 0x510100, opc: 0, slot_kind: [0, 1], slot_val: [0x11, 0x64], long_size: 8, out_size: 4, offset: 0, extra_kind: 0 },
        CaseSpec { name: "add_const_free", base: 0x510200, opc: 0, slot_kind: [1, 0], slot_val: [0x64, 0x22], long_size: 8, out_size: 4, offset: 0, extra_kind: 0 },
        CaseSpec { name: "add_const_const", base: 0x510300, opc: 0, slot_kind: [1, 1], slot_val: [0x64, 0xc8], long_size: 8, out_size: 4, offset: 0, extra_kind: 0 },
        // Other commuting opcodes with free tails.
        CaseSpec { name: "mult_free_free", base: 0x510400, opc: 1, slot_kind: [0, 0], slot_val: [7, 6], long_size: 8, out_size: 4, offset: 0, extra_kind: 0 },
        CaseSpec { name: "and_free_free", base: 0x510500, opc: 2, slot_kind: [0, 0], slot_val: [0xf0, 0x0f], long_size: 8, out_size: 2, offset: 0, extra_kind: 0 },
        CaseSpec { name: "or_free16", base: 0x510600, opc: 3, slot_kind: [0, 0], slot_val: [0xa1, 0xb2], long_size: 16, out_size: 8, offset: 0, extra_kind: 0 },
        // Single-input arm.
        CaseSpec { name: "negate_free", base: 0x510700, opc: 5, slot_kind: [0, 0], slot_val: [0x33, 0], long_size: 8, out_size: 4, offset: 0, extra_kind: 0 },
        // Bitwise commutes at a non-zero subpiece offset.
        CaseSpec { name: "xor_free_off4", base: 0x510800, opc: 4, slot_kind: [0, 0], slot_val: [0x44, 0x55], long_size: 8, out_size: 4, offset: 4, extra_kind: 0 },
        CaseSpec { name: "mult_free_const", base: 0x510900, opc: 1, slot_kind: [0, 1], slot_val: [9, 3], long_size: 8, out_size: 4, offset: 0, extra_kind: 0 },
        // dup-reuse leg: the SAME written vn feeds both slots — only one
        // new SUBPIECE is created (cc:4636 guard, cc:4646 reuse).
        CaseSpec { name: "xor_same_written", base: 0x510a00, opc: 4, slot_kind: [4, 5], slot_val: [0x66, 0x66], long_size: 8, out_size: 4, offset: 0, extra_kind: 0 },
        // cc:4623-4628 reject: outvn feeds a ZEXT back to insize.
        CaseSpec { name: "add_free_zext_overlap", base: 0x510b00, opc: 0, slot_kind: [0, 0], slot_val: [0x11, 0x22], long_size: 8, out_size: 4, offset: 0, extra_kind: 1 },
        // cc:4621 reject: a second SUBPIECE also reads base.
        CaseSpec { name: "add_two_readers", base: 0x510c00, opc: 0, slot_kind: [0, 0], slot_val: [0x11, 0x22], long_size: 8, out_size: 4, offset: 0, extra_kind: 2 },
    ];

    let mode = std::env::args().nth(1).unwrap_or_default();
    let run_cases: Vec<CaseSpec> = if mode == "trap_dupfree" {
        vec![CaseSpec { name: "trap_dupfree", base: 0x520000, opc: 4, slot_kind: [0, 3], slot_val: [0x77, 0x77], long_size: 8, out_size: 4, offset: 0, extra_kind: 0 }]
    } else if mode == "normal" {
        cases
    } else {
        eprintln!("unknown mode: {mode}");
        std::process::exit(2);
    };

    for spec in run_cases {
        let base = spec.base;
        let built = tail_commute_case(
            &mut fd,
            &block,
            opc_of(&spec),
            base,
            spec.slot_kind,
            spec.slot_val,
            spec.long_size,
            spec.out_size,
            spec.offset,
            spec.extra_kind,
        );
        let sub_commute = RuleSubCommute::new();
        let apply = sub_commute
            .apply_op(&built.sub_op.0, &mut fd)
            .expect("subcommute apply");
        println!("case={}|sub_apply={apply}", spec.name);
        dump_case_window(&fd, base, base + 0x80);
        dump_keep_lines(&fd, &built.keeps, &built.tags, base, base + 0x80);
        println!("endcase");
    }
}

fn main() {
    run();
}
